libslide/src/partial_evaluator/flatten.rs

```diff
@@ -155,6 +172,318 @@ fn flatten_add_or_sub(o_lhs: &Rc<Expr>, o_rhs: &Rc<Expr>, is_subtract: bool) ->
+/// ```
+///
+/// The next two terms are added to the registry and constant-folded, respectively.
+///
+/// ```text
+/// m           div                  registry       const
+/// -||----------------------
+/// ||| (/ 1 (* x (/ 1 y))) |      {x: 1, y: -1}     2/5
+/// -||----------------------
+/// ```
+///
+/// Now we see a division `A` on the division side. This is the same thing as multiplying the
+/// reciprocal of `A` on the top level!
+///
+/// > Let's break down a simpler example. Observe that `1 / (2 / 3)` is equivalent to `3 / 2`. The
+/// > flattening list for `1 / (2 / 3)` after the folding of `1` looks like
+/// >
+/// > ```text
+/// > m     div       const
+/// > -||----------
+/// > ||| (/ 2 3) |     1
+/// > -||----------
+/// > ```
+/// >
+/// > Now we simply add the reciprocal of the division expression to the multiplication side.
+/// >
+/// > ```text
+/// >     mul     d   const
+/// > ----------||-
+/// > | (/ 3 2) |||     1
+/// > ----------||-
+/// > ```
+/// >
+/// > And we already know this gets unfolded as
+/// >
+/// > ```text
+/// >  mul  div   const
+/// > ----||----
+/// > | 3 || 2 |    1
+/// > ----||----
+/// > ```
+/// >
+/// > So we can skip adding the entire division to the multiplication side, instead adding the
+/// > operands where appropriate. The rest of the constant folding follows trivially.
+///
+/// Back to the original example, whose current state is
+///
+/// ```text
+/// m           div                  registry       const
+/// -||----------------------
+/// ||| (/ 1 (* x (/ 1 y))) |      {x: 1, y: -1}     2/5
+/// -||----------------------
+/// ```
+///
+/// Let's apply our "division in division" algorithm: the left operand gets divided, and the right
+/// operand gets multiplied.
+///
+/// ```text
+///       mul         div         registry       const
+/// ----------------||----
+/// | (* x (/ 1 y)) || 1 |      {x: 1, y: -1}     2/5
+/// ----------------||----
+/// ```
+///
+/// Now we unfold the new expressions on the multiplication side.
+///
+/// ```text
+///      mul        div         registry       const
+/// --------------||----
+/// | x | (/ 1 y) || 1 |      {x: 1, y: -1}     2/5
+/// --------------||----
+/// ```
+///
+/// Two steps this time:
+///
+/// ```text
+///  mul    div           registry       const
+/// ----||--------
+/// | 1 || 1 | y |      {x: 2, y: -1}     2/5
+/// ----||--------
+/// ```
+///
+/// Three steps this time:
+///
+/// ```text
+/// m  d        registry       const
+/// -||-
+/// ||||      {x: 2, y: -2}     2/5
+/// -||-
+/// ```
+///
+/// And now, all that needs to be done is to construct the flattened expression `2/5 * x^2 / y^-2`.
+fn flatten_mul_or_div(o_lhs: &Rc<Expr>, o_rhs: &Rc<Expr>, is_div: bool) -> Rc<Expr> {
+    let lhs = flatten_expr(o_lhs);
+    let rhs = flatten_expr(o_rhs);
+
+    let mut coeff = 1.;
+    // Term -> # of times it is multiplied. Negative values are equivalent to division.
+    let mut terms = BTreeMap::<&Rc<Expr>, f64>::new();
+
```

Comment:
so why are you using a BTreeMap. I don't see a first or last accesssor here which is my understanding as the main advantage. Wouldn't caching in hashmaps be better

Reply:
HashMap's ordering is non-deterministic, whereas BTreeMap is always sorted. So if we use a HashMap the order of terms in the final expression may not always be the same, which for the system isn't a big a deal but for unit tests matter. LMK if you would prefer I change to a HashMap and call `normalize` during the unit tests.

Reply:
ah I see. I think BTreeMap is fine then bc I'm lazy